libyul/backends/evm/ControlFlowGraphBuilder.cpp

```diff
@@ -0,0 +1,484 @@
+	m_currentBlock = nullptr;
+}
+
+void ControlFlowGraphBuilder::jump(CFG::BasicBlock& _target, bool backwards)
+{
+	yulAssert(m_currentBlock, "");
+	m_currentBlock->exit = CFG::BasicBlock::Jump{&_target, backwards};
+	_target.entries.emplace_back(m_currentBlock);
+	m_currentBlock = &_target;
+}
+
+void ControlFlowGraphBuilder::operator()(If const& _if)
+{
+	auto&& [ifBranch, afterIf] = makeConditionalJump(std::visit(*this, *_if.condition));
+	m_currentBlock = ifBranch;
+	(*this)(_if.body);
+	jump(*afterIf);
+}
+
+void ControlFlowGraphBuilder::operator()(Switch const& _switch)
+{
+	yulAssert(m_currentBlock, "");
+	auto ghostVariableId = m_graph.ghostVariables.size();
+	YulString ghostVariableName("GHOST[" + to_string(ghostVariableId) + "]");
+	auto& ghostVar = m_graph.ghostVariables.emplace_back(Scope::Variable{""_yulstring, ghostVariableName});
+
+	// Artificially generate:
+	// let <ghostVariable> := <switchExpression>
+	VariableSlot ghostVarSlot{ghostVar, debugDataOf(*_switch.expression)};
+	m_currentBlock->operations.emplace_back(CFG::Operation{
+		Stack{std::visit(*this, *_switch.expression)},
+		Stack{ghostVarSlot},
+		CFG::Assignment{_switch.debugData, {ghostVarSlot}}
+	});
+
+	BuiltinFunction const* equalityBuiltin = m_dialect.equalityFunction({});
+	yulAssert(equalityBuiltin, "");
+
+	// Artificially generate:
+	// eq(<literal>, <ghostVariable>)
+	auto makeValueCompare = [&](Literal const& _value) {
+		yul::FunctionCall const& ghostCall = m_graph.ghostCalls.emplace_back(yul::FunctionCall{
+			_value.debugData,
+			yul::Identifier{{}, "eq"_yulstring},
+			{_value, Identifier{{}, ghostVariableName}}
+		});
+		CFG::Operation& operation = m_currentBlock->operations.emplace_back(CFG::Operation{
+			Stack{ghostVarSlot, LiteralSlot{valueOfLiteral(_value), _value.debugData}},
+			Stack{TemporarySlot{ghostCall, 0}},
+			CFG::BuiltinCall{_switch.debugData, *equalityBuiltin, ghostCall, 2},
+		});
+		return operation.output.front();
+	};
+	CFG::BasicBlock& afterSwitch = m_graph.makeBlock();
+	for (auto const& switchCase: _switch.cases | ranges::views::drop_last(1))
+	{
+		yulAssert(switchCase.value, "");
+		auto&& [caseBranch, elseBranch] = makeConditionalJump(makeValueCompare(*switchCase.value));
+		m_currentBlock = caseBranch;
+		(*this)(switchCase.body);
+		jump(afterSwitch);
+		m_currentBlock = elseBranch;
+	}
+	Case const& switchCase = _switch.cases.back();
+	if (switchCase.value)
+	{
+		CFG::BasicBlock& caseBranch = m_graph.makeBlock();
+		makeConditionalJump(makeValueCompare(*switchCase.value), caseBranch, afterSwitch);
+		m_currentBlock = &caseBranch;
+		(*this)(switchCase.body);
+	}
+	else
+		(*this)(switchCase.body);
+	jump(afterSwitch);
+}
+
+void ControlFlowGraphBuilder::operator()(ForLoop const& _loop)
+{
+	ScopedSaveAndRestore scopeRestore(m_scope, m_info.scopes.at(&_loop.pre).get());
+	(*this)(_loop.pre);
+
+	std::optional<bool> constantCondition;
+	if (auto const* literalCondition = get_if<yul::Literal>(_loop.condition.get()))
```

Comment:
I thought about it, but it's not that easy. You'll end up with more jumps in the graph, so to get the same results you'd then need to compress them later... which gets messy with simultanously tracking backwards jumps. Could probably be done, but not sure it's worth it.
In general the question is if we want to require to pre-transform the AST or run code generation on arbitrary ASTs. If the benefits of pre-transforming are minor, I'd say it's nicer to be able to run on arbitrary ASTs.
I did write a sketch of a version that goes further and even runs the expression splitter to avoid needing the ``TemporarySlot`` things altogether (it's https://github.com/ethereum/solidity/tree/newCodeGeneratorWithExpressionSplitter, but not sure how up to date that branch is) - but in the end wasn't convinced that it's really easier and beneficial (in fact the way the new code generator is currently written it does rely on, resp. benefit from the fact that statements are merged, if merging is compatible with the execution order).